### PR TITLE
Add GMM cookbook page

### PR DESCRIPTION
**Pull request by sanuj**
@karlnapf Please review. A few doubts:
1. I have added this in clustering dir. Shall I add it in classification?
2. Currently there is no output for integration tests because all the values change every time GMM is run. Any way around this?

### CODE REVIEW

**Comment by sanuj** [comment]
I think [this member function](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/clustering/GMM.h#L68) has a confusing name. It sets the training features but its name is `train`.

**Comment by karlnapf** [comment]
Clustering dir is just what it should be :)

**Comment by karlnapf** [comment]
No output for the test should be ok, as this can be covered by unit tests

**Comment by karlnapf** [comment]
If you want, you can make something like extract the cluster centres and variances. But this will probably change from run to run. Need to try

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+A Gaussian Mixture Model is a probabilistic model that assumes all the data points are generated from a mixture of a finite number of Gaussian distributions with unknown parameters.
+
+.. math::
+
+    p(\theta|\boldsymbol{X}) = \sum_{i=1}^{K}{\phi_i \cal{N}(\boldsymbol{\mu_i}, \boldsymbol{\Sigma_i})}
+
+where :math:`p(\theta|\boldsymbol{X})` is posterior distribution, :math:`K` denotes number of mixture components, :math:`\phi_i` denotes weight for :math:`i^{th}` component, :math:`\cal{N}` denotes normal distribution with mean vector :math:`\boldsymbol{\mu_i}` and covariance matrix :math:`\boldsymbol{\Sigma_i}`.
+
+See Chapter 20 in :cite:`barber2012bayesian` for a detailed introduction.
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Mention that this implementation uses the EM algorithm which is executed until some convergence criterion (which) is met

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+Start by creating CDenseFeatures (here 64 bit floats aka RealFeatures) as
+
+.. sgexample:: gmm.sg:create_features
+
+Initialize :sgclass:`GMM` by deciding the number of components for the mixture model.
+
+.. sgexample:: gmm.sg:create_gmm_instance
+
+Provide training features to the :sgclass:`GMM` object and train it by using EM algorithm. SMEM algorithm can also be used for training. After training, one can sample data-points from the trained model.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Need to mention / reference what SMEM is

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+.. sgexample:: gmm.sg:train_sample
+
+Extract parameters like :math:`\phi_i`, :math:`\boldsymbol{\mu_i}` and :math:`\boldsymbol{\Sigma_i}` for any componenet from the trained model.
+
+.. sgexample:: gmm.sg:extract_params
+
+Obtain log likelihood of belonging to clusters and being generated by this model which can be used for classification.
+
+.. sgexample:: gmm.sg:cluster_output
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: What is also needed to to extract learned parameters

File: examples/meta/src/clustering/gmm.sg [hunk 4]
```diff
@@ -0,0 +1,28 @@
+CSVFile f_feats_train("../../data/classifier_binary_2d_linear_features_train.dat")
```
Comment (karlnapf) [inline examples/meta/src/clustering/gmm.sg]: Can you maybe use a multiclass dataset here?

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+Gaussian Mixture Models
+=======================
+
+A Gaussian Mixture Model is a probabilistic model that assumes all the data points are generated from a mixture of a finite number of Gaussian distributions with unknown parameters.
+
+.. math::
+
+    p(\theta|\boldsymbol{X}) = \sum_{i=1}^{K}{\phi_i \cal{N}(\boldsymbol{\mu_i}, \boldsymbol{\Sigma_i})}
+
+where :math:`p(\theta|\boldsymbol{X})` is posterior distribution, :math:`K` denotes number of mixture components, :math:`\phi_i` denotes weight for :math:`i^{th}` component, :math:`\cal{N}` denotes normal distribution with mean vector :math:`\boldsymbol{\mu_i}` and covariance matrix :math:`\boldsymbol{\Sigma_i}`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: I wouldnt use \boldsymbol   but it is fine if you want that

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: I wouldnt use \boldsymbol   but it is fine if you want that

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Never put plain text in math mode.....
:math i^{th}

The "th" can go after the math

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Never put plain text in math mode.....
:math i^{th}

The "th" can go after the math

**Comment by vigsterkr** [comment]
do we know why c# and java modular tests for all the generated examples fails?

File: doc/cookbook/source/examples/clustering/gmm.rst (hunk 1 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: ":math:`\cal{N}` denotes normal distribution"

put N(\mu,\Sigma) denote a multivariate normal .....

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+Initialize :sgclass:`GMM` by deciding the number of components for the mixture model.
+
+.. sgexample:: gmm.sg:create_gmm_instance
+
+Provide training features to the :sgclass:`GMM` object and train it by using EM algorithm. SMEM algorithm can also be used for training. After training, one can sample data-points from the trained model.
+
+.. sgexample:: gmm.sg:train_sample
+
+Extract parameters like :math:`\phi_i`, :math:`\boldsymbol{\mu_i}` and :math:`\boldsymbol{\Sigma_i}` for any componenet from the trained model.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: In the code, you extract all \phis, not just phi_i

**Comment by karlnapf** [comment]
@vigsterkr I wasnt aware that these are failing. They were not a while ago

**Comment by karlnapf** [comment]
@sanuj I think you are responsible for the cookbook build failure:
https://travis-ci.org/shogun-toolbox/shogun/jobs/132213321#L2585

**Comment by karlnapf** [comment]
We might need to change the grammar of csharp and java meta examples. See `targets/*.json`
Afaik, csharp and java numerical types are always translated to Double. See what I mean?
Let me know if you need help fixing this

File: examples/meta/src/clustering/gmm.sg [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+#![extract_params]
+Real component_num = 1
+RealVector nth_mean = gmm.get_nth_mean(component_num)
+RealMatrix nth_cov = gmm.get_nth_cov(component_num)
+RealVector coef = gmm.get_coef()
+#![extract_params]
+
+#![cluster_output]
+RealVector any_point = nth_mean
```
Comment (karlnapf) [inline examples/meta/src/clustering/gmm.sg]: no need for that, all Real* variables are stored anyways

File: src/shogun/lib/common.h [hunk 8]
```diff
@@ -56,7 +56,9 @@ typedef long double floatmax_t;
 #define STRING_LEN_STR             "256"
 typedef char                       string_t[STRING_LEN];
 
-typedef int                        machine_int_t;
+typedef int                        int32_t;
```
Comment (karlnapf) [inline src/shogun/lib/common.h]: this really should be in a seperate patch btw

**Comment by OXPHOS** [comment]
I guess I'm experiencing the same problem..

**Comment by karlnapf** [comment]
Checking locally

File: examples/meta/src/clustering/gmm.sg [hunk 9]
```diff
@@ -0,0 +1,27 @@
+CSVFile f_feats_train("../../data/classifier_4class_2d_linear_features_train.dat")
+
+#![create_features]
+RealFeatures features_train(f_feats_train)
+#![create_features]
+
+#![create_gmm_instance]
+real num_components = 3
```
Comment (karlnapf) [inline examples/meta/src/clustering/gmm.sg]: dude, it is this line
has to be "int"

**Comment by karlnapf** [comment]
Works on my machine using "int" for num_compoenents

**Comment by sanuj** [comment]
@karlnapf can we merge this?

**Comment by karlnapf** [comment]
Nope, it needs integration testing data, see cookbook readme

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 10]
```diff
@@ -0,0 +1,49 @@
+=======================
+Gaussian Mixture Models
+=======================
+
+A Gaussian Mixture Model is a probabilistic model that assumes all the data points are generated from a mixture of a finite number of Gaussian distributions with unknown parameters.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: "that assumes that data are generated from a finite mixutre of Gaussians with unknown parameters" is better

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+.. math::
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: The model likelihood can be written as

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    p(\theta|\boldsymbol{X}) = \sum_{i=1}^{K}{\phi_i \cal{N}(\boldsymbol{\mu_i}, \boldsymbol{\Sigma_i})}
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: This should be p(x|theta) (and you should define \theta below

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: for the Gaussian N, please write it as N(x | \mu_i, \Sigma_i)

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: can we use \pi_i for the weights?

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+where :math:`p(\theta|\boldsymbol{X})` is posterior distribution, :math:`K` denotes number of mixture components, :math:`\phi_i` denotes weight for :math:`i`-th component, :math:`\cal{N}` denotes a multivariate normal distribution with mean vector :math:`\boldsymbol{\mu_i}` and covariance matrix :math:`\boldsymbol{\Sigma_i}`.
+
+Expectation Maximization(EM) algorithm is used to train the model iteratively until the local maxima in the model's likelihood is obtained.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: "The expectation maximization (EM) algorithm is used"

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+See Chapter 20 in :cite:`barber2012bayesian` for a detailed introduction.
+
+-------
+Example
+-------
+
+Start by creating CDenseFeatures (here 64 bit floats aka RealFeatures) as
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: We start by

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+.. sgexample:: gmm.sg:create_features
+
+Initialize :sgclass:`GMM` by deciding the number of components for the mixture model.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: We initialize

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+.. sgexample:: gmm.sg:create_gmm_instance
+
+Provide training features to the :sgclass:`GMM` object and train it by using EM algorithm. Split-Merge Expectation-Maximization algorithm :cite:`ueda2000smem` can also be used for training which can be found in :sgclass:`GMM` documentation. After training, one can sample data-points from the trained model.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: All your sentences are missing the subject ;)
"We provide training features"

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Please mention the split merge version at the end, the first part should just be standard EM

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+.. sgexample:: gmm.sg:train_sample
+
+Extract parameters like :math:`\Phi`, :math:`\boldsymbol{\mu_i}` and :math:`\boldsymbol{\Sigma_i}` for any componenet from the trained model.
+
+.. sgexample:: gmm.sg:extract_params
+
+Obtain log likelihood of belonging to clusters and being generated by this model which can be used for classification.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: How can it be used for classification? 
Either show or remove

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 18]
```diff
@@ -0,0 +1,53 @@
+=======================
+Gaussian Mixture Models
+=======================
+
+A Gaussian Mixture Model is a probabilistic model that assumes that data are generated from a finite mixture of Gaussians with unknown parameters. The model likelihood can be written as:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: mixture model (Only names are capital in English, like "Gaussian")

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+.. math::
+
+    p(x|\theta) = \sum_{i=1}^{K}{\pi_i N(x|\mu_i, \Sigma_i)}
+
+where :math:`p(x|\theta)` is probability distribution given :math:`\theta:=\{\mu_i, \Sigma_i\}_{i=1}^K`, :math:`K` denotes number of mixture components, :math:`\pi_i` denotes weight for :math:`i`-th component, :math:`N` denotes a multivariate normal distribution with mean vector :math:`\mu_i` and covariance matrix :math:`\Sigma_i`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: theta includes the weights \pi_i as well

Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Can you use `\mathcal{N}` for the Gaussian (maybe misunderstood me earlier, I just wanted the x in there)

File: doc/cookbook/source/examples/clustering/gmm.rst [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+The expectation maximization (EM) algorithm is used to learn parameters of the model, via finding a local maximum of the likelihood.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/clustering/gmm.rst]: Sorry, my mistake, it should be
"... via finding a local maximum of a lower bound on the likelihood"

File: doc/cookbook/source/references.bib [hunk 21]
```diff
@@ -24,3 +24,13 @@ @book{Rasmussen2005GPM
   year = {2005},
   publisher = {The MIT Press}
 }
+@article{ueda2000smem,
+  title={SMEM Algorithm for Mixture Models},
+  author={N. Ueda and R. Nakano and Z. Ghahramani and G.E.Hinton},
```
Comment (karlnapf) [inline doc/cookbook/source/references.bib]: minor: space after "G.E." missing (surname is not separated)

**Comment by karlnapf** [comment]
The only thing really missing is integration testing data, can you add? Then we can merge

**Comment by sanuj** [comment]
@karlnapf Integration test data keeps changing every-time the tests are run, so the tests will fail.

**Comment by karlnapf** [comment]
Ah yes, sorry I forgot

**Comment by karlnapf** [comment]
In that case, lets merge :)

**Pull request merged**